pkg/services/secrets/kvstore/migrations/migrator.go

```diff
@@ -9,41 +9,58 @@ import (
 	"github.com/grafana/grafana/pkg/infra/serverlock"
 	datasources "github.com/grafana/grafana/pkg/services/datasources/service"
 	"github.com/grafana/grafana/pkg/services/secrets/kvstore"
+	"github.com/grafana/grafana/pkg/setting"
 )
 
 var logger = log.New("secret.migration")
 
+const actionName = "secret migration task "
+
 // SecretMigrationService is used to migrate legacy secrets to new unified secrets.
 type SecretMigrationService interface {
 	Migrate(ctx context.Context) error
 }
 
 type SecretMigrationServiceImpl struct {
-	Services          []SecretMigrationService
-	ServerLockService *serverlock.ServerLockService
+	services                 []SecretMigrationService
+	ServerLockService        *serverlock.ServerLockService
+	migrateToPluginService   *kvstore.MigrateToPluginService
+	migrateFromPluginService *kvstore.MigrateFromPluginService
 }
 
 func ProvideSecretMigrationService(
+	cfg *setting.Cfg,
 	serverLockService *serverlock.ServerLockService,
 	dataSourceSecretMigrationService *datasources.DataSourceSecretMigrationService,
-	pluginSecretMigrationService *kvstore.PluginSecretMigrationService,
+	migrateToPluginService *kvstore.MigrateToPluginService,
+	migrateFromPluginService *kvstore.MigrateFromPluginService,
 ) *SecretMigrationServiceImpl {
 	services := make([]SecretMigrationService, 0)
 	services = append(services, dataSourceSecretMigrationService)
-	// pluginMigrationService should always be the last one
-	services = append(services, pluginSecretMigrationService)
+	// Plugin migration should always be last; should either migrate to or from, not both
+	// This is because the migrateTo checks for use_plugin = true, in which case we should always
+	// migrate by default to ensure users don't lose access to secrets. If migration has
+	// already occurred, the migrateTo function will be called but it won't do anything
+	if cfg.SectionWithEnvOverrides("secrets").Key("migrate_from_plugin").MustBool(false) {
```

Comment:
What happens if use_plugin is false but migrate_from_plugin is true? Does the plugin loads and migration happens normally?

Reply:
I will test this case but yes, it would do the backwards migration then stop the plugin again. Now that I say it, I am not sure if that is a case we actually want to support. Perhaps this needs to be guarded with a check for use_plugin